### DISCUSSION TITLE
Interrupt from PMU counter

### DISCUSSION THREAD

**Discussion by Everaldo-Gomes** [bao-project/bao-hypervisor#152] [General]
Hi,

I've configured the PMU to count an event. 
The interrupt for that counter is enabled.
When the counter overflows, it's setting the overflow bit.

My goal is to generate an interrupt when that counter overflows.

In the [interrupts.c](https://github.com/bao-project/bao-hypervisor/blob/main/src/core/interrupts.c) I'm using the `interrupts_init` to set my interrupt
```
inline void interrupts_init()
{
    interrupts_arch_init();
    
    if (cpu()->id == CPU_MASTER) {
        interrupts_reserve(IPI_CPU_MSG, cpu_msg_handler);
	interrupts_reserve(MY_IRQ_ID, my_pmu_overflow_handler);  <--- added
    }

    interrupts_cpu_enable(IPI_CPU_MSG, true);
    interrupts_cpu_enable(MY_IRQ_ID, true);  <--- added		
}
```

By doing so,  I believed I would be able to get my function called in the `interrupts_handle`

```
enum irq_res interrupts_handle(irqid_t int_id)
{
    //---------------- Added ---------------------
     if (int_id == MY_IRQ_ID)
     {
        my_pmu_overflow_handler();
        return HANDLED_BY_HYP;
     }
    //-----------------------------------------------
    
    else if (vm_has_interrupt(cpu()->vcpu->vm, int_id)) {
        vcpu_inject_hw_irq(cpu()->vcpu, int_id);

        return FORWARD_TO_VM;

    } else if (interrupt_assigned_to_hyp(int_id)) {
        interrupt_handlers[int_id](int_id);

        return HANDLED_BY_HYP;

    } else {
        ERROR("received unknown interrupt id = %d", int_id);
    }
}
```

But not even the `irq_res interrupts_handle` is being called. Am I forgetting something?
And how do I know if overflow interrupt will have the ID I want?

Platform: Raspberry Pi 4b
Demo: Dual-guest Linux+FreeRTOS

**Comment by Diogo21Costa** [comment]
Hi @Everaldo-Gomes,

First of all, sorry for the delay in my response. You can achieve this without modifying the interrupts.c file. Please take a look at this [fork](https://github.com/AFOliveira/bao-hypervisor-mem-throttling) of Bao. It shows a simple implementation of a memory throttling mechanism at the hypervisor level, which uses PMU interrupts to control the budget of memory accesses and a generic timer to reestablish the budget assigned to each CPU.

Here are a few points to guide you through the repository:

- In `/src/core/mem_throt.c`, you can find the implementation of the mechanism, including the timer initialization and PMU events configuration.
- In `/src/core/inc/events.h`, you'll find utilities to configure the PMU events, which will redirect you to the corresponding platform architecture (for the Raspberry Pi 4b, this would be the ARMv8).
- In `/src/arch/armv8/pmu.c`, you can find the PMU configuration, including the allocation and enabling of PMU counters as well as the interrupt assignment. This part might be of particular interest to you.

Currently, this repo is a work-in-progress and presents a very basic implementation of memory throttling. It is not yet part of the official Bao hypervisor and has only been tested on the Xilinx Ultrascale+ ZCU104. To use it on the Raspberry Pi, you will need to update the [platform description file](https://github.com/AFOliveira/bao-hypervisor-mem-throttling/blob/main/src/platform/rpi4/rpi4_desc.c). You can take a look at the ZCU104 example [here](https://github.com/AFOliveira/bao-hypervisor-mem-throttling/blob/main/src/platform/zcu104/zcu104_desc.c).

For the overflow interrupt ID, you'll need to consult the [technical reference manual](https://datasheets.raspberrypi.com/bcm2711/bcm2711-peripherals.pdf) of the platform. You can find it here. On page 89, Figure 7 shows that the Core 0 PMU IRQ ID is 48, which is the ID you need to set in the platform description file.

Please feel free to reach out if you need further assistance.

**Reply by Everaldo-Gomes** [reply]
It's working now. Thank you so much!!
